### ISSUE TITLE
IndexOutofBoundsException when creating pulse because of remapped column

### ISSUE THREAD

**Issue by landreville** [metabase/metabase#11844]
**Describe the bug**
When creating a pulse there is an IndexOutOfBoundsException generated.

**Logs**
```
GET /api/pulse/preview_card_info/181 500 334.6 ms (21 DB calls) 
{:message nil,
 :type java.lang.IndexOutOfBoundsException,
 :stacktrace
 ("clojure.lang.RT.nthFrom(RT.java:928)"
  "clojure.lang.RT.nth(RT.java:897)"
  "--> pulse.render.body$query_results__GT_header_row$iter__32026__32030$fn__32031.invoke(body.clj:84)"
  "pulse.render.table$render_table_head$iter__31874__31878$fn__31879.invoke(table.clj:59)"
  "api.pulse$fn__63391.invokeStatic(pulse.clj:166)"
  "api.pulse$fn__63391.invoke(pulse.clj:159)"
  "middleware.auth$enforce_authentication$fn__66128.invoke(auth.clj:14)"
  "routes$fn__67481$fn__67482.doInvoke(routes.clj:56)"
  "middleware.exceptions$catch_uncaught_exceptions$fn__66227.invoke(exceptions.clj:104)"
  "middleware.exceptions$catch_api_exceptions$fn__66224.invoke(exceptions.clj:92)"
  "middleware.log$log_api_call$fn__67857$fn__67858.invoke(log.clj:178)"
  "middleware.log$log_api_call$fn__67857.invoke(log.clj:172)"
  "middleware.security$add_security_headers$fn__66190.invoke(security.clj:121)"
  "middleware.json$wrap_json_body$fn__67562.invoke(json.clj:64)"
  "middleware.json$wrap_streamed_json_response$fn__67580.invoke(json.clj:100)"
  "middleware.session$bind_current_user$fn__62721$fn__62722.invoke(session.clj:209)"
  "middleware.session$do_with_current_user.invokeStatic(session.clj:184)"
  "middleware.session$do_with_current_user.invoke(session.clj:177)"
  "middleware.session$bind_current_user$fn__62721.invoke(session.clj:208)"
  "middleware.session$wrap_current_user_id$fn__62706.invoke(session.clj:161)"
  "middleware.session$wrap_session_id$fn__62650.invoke(session.clj:125)"
  "middleware.auth$wrap_api_key$fn__66136.invoke(auth.clj:27)"
  "middleware.misc$maybe_set_site_url$fn__67884.invoke(misc.clj:56)"
  "middleware.misc$bind_user_locale$fn__67887.invoke(misc.clj:72)"
  "middleware.misc$add_content_type$fn__67872.invoke(misc.clj:28)"
  "middleware.misc$disable_streaming_buffering$fn__67895.invoke(misc.clj:87)")}
```

**To Reproduce**
1. Click on create new pulse
2. Select a question under Pick your data
3. Loading symbol animated on the question
4. Logs (above) appear in Javascript console and server log

So far I haven't managed to re-produce this on the sample dataset.

**Expected behavior**
A preview of the selected question is displayed.

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:72.0) Gecko/20100101 Firefox/72.0",
    "vendor": ""
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_232-b09",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_232",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.232-b09",
    "os.name": "Linux",
    "os.version": "3.10.0-1062.4.1.el7.x86_64",
    "user.language": "en",
    "user.timezone": "America/Toronto"
  },
  "metabase-info": {
    "databases": [
      "googleanalytics",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "run-mode": "prod",
    "version": {
      "tag": "v0.34.1",
      "date": "2020-01-13",
      "branch": "release-0.34.x",
      "hash": "265695c"
    },
    "settings": {
      "report-timezone": "US/Eastern"
    }
  }
}
```

**Severity**
It appears that pulses cannot be created and there is no workaround.

**Additional context**

The dataset I'm using is a Postgres database. The question uses one fact table and a couple dimension tables. The foreign keys to dimension tables are setup to display a foreign key value in the results. 

The question is a table of results and renders fine as a question by itself. 

There are two filters on the question. One is a date filter for current month and one is a boolean filter on a dimension column.

**Comment by landreville** [comment]
I've managed to reproduce this on the sample data set.

Here is the data model setup (using foreign key values for the display value):

![image](https://user-images.githubusercontent.com/24076/73684767-d4591980-4692-11ea-8969-511bec15ae52.png)


Here's the question:
![image](https://user-images.githubusercontent.com/24076/73684704-bab7d200-4692-11ea-8040-ebbb5e1d1542.png)

And the log output on the sample dataset is the same:

```
[6187d89d-9df9-43e7-b79c-481918cd46d7] 2020-02-03T14:36:34-05:00 ERROR metabase.middleware.log GET /api/pulse/preview_card_info/6 500 1.3 s (22 DB calls) 
{:message nil,
 :type java.lang.IndexOutOfBoundsException,
 :stacktrace
 ("clojure.lang.RT.nthFrom(RT.java:928)"
  "clojure.lang.RT.nth(RT.java:897)"
  "--> pulse.render.body$query_results__GT_header_row$iter__32793__32797$fn__32798.invoke(body.clj:71)"
  "pulse.render.table$render_table_head$iter__32650__32654$fn__32655.invoke(table.clj:59)"
  "api.pulse$fn__65440.invokeStatic(pulse.clj:166)"
  "api.pulse$fn__65440.invoke(pulse.clj:159)"
  "middleware.auth$enforce_authentication$fn__48819.invoke(auth.clj:14)"
  "routes$fn__67566$fn__67567.doInvoke(routes.clj:56)"
  "middleware.exceptions$catch_uncaught_exceptions$fn__48775.invoke(exceptions.clj:104)"
  "middleware.exceptions$catch_api_exceptions$fn__48772.invoke(exceptions.clj:92)"
  "middleware.log$log_api_call$fn__48685$fn__48686.invoke(log.clj:178)"
  "middleware.log$log_api_call$fn__48685.invoke(log.clj:172)"
  "middleware.security$add_security_headers$fn__48736.invoke(security.clj:121)"
  "middleware.json$wrap_json_body$fn__48904.invoke(json.clj:64)"
  "middleware.json$wrap_streamed_json_response$fn__48922.invoke(json.clj:100)"
  "middleware.session$bind_current_user$fn__47354$fn__47355.invoke(session.clj:209)"
  "middleware.session$do_with_current_user.invokeStatic(session.clj:184)"
  "middleware.session$do_with_current_user.invoke(session.clj:177)"
  "middleware.session$bind_current_user$fn__47354.invoke(session.clj:208)"
  "middleware.session$wrap_current_user_id$fn__47339.invoke(session.clj:161)"
  "middleware.session$wrap_session_id$fn__47283.invoke(session.clj:125)"
  "middleware.auth$wrap_api_key$fn__48827.invoke(auth.clj:27)"
  "middleware.misc$maybe_set_site_url$fn__48800.invoke(misc.clj:56)"
  "middleware.misc$bind_user_locale$fn__48803.invoke(misc.clj:72)"
  "middleware.misc$add_content_type$fn__48788.invoke(misc.clj:28)"
  "middleware.misc$disable_streaming_buffering$fn__48811.invoke(misc.clj:87)")}

```

**Comment by nickhamlin** [comment]
We're running into the same thing, but with a slightly different error message. 

Install details (we're running metabase in docker and connecting to MySQL data sources):
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.4+11",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.4",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.4+11",
    "os.name": "Linux",
    "os.version": "4.9.41-moby",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mysql",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "run-mode": "prod",
    "version": {
      "date": "2019-09-20",
      "tag": "v0.33.3",
      "branch": "release-0.33.x",
      "hash": "7d81ff4"
    },
    "settings": {
      "report-timezone": "US/Eastern"
    }
  }
}

Error message
```
{:message nil,
:type java.lang.IndexOutOfBoundsException,
:stacktrace
("clojure.lang.RT.nthFrom(RT.java:928)"
"clojure.lang.RT.nth(RT.java:897)"
"--> pulse.render.body$query_results__GT_header_row$iter__28668__28672$fn__28673.invoke(body.clj:71)"
"pulse.render.table$render_table_head$iter__28549__28553$fn__28554.invoke(table.clj:52)"
"api.pulse$fn__61055.invokeStatic(pulse.clj:165)"
"api.pulse$fn__61055.invoke(pulse.clj:158)"
"middleware.auth$enforce_authentication$fn__63874.invoke(auth.clj:14)"
"routes$fn__65248$fn__65249.doInvoke(routes.clj:56)"
"middleware.exceptions$catch_uncaught_exceptions$fn__63973.invoke(exceptions.clj:104)"
"middleware.exceptions$catch_api_exceptions$fn__63970.invoke(exceptions.clj:92)"
"middleware.log$log_api_call$fn__65622$fn__65623.invoke(log.clj:170)"
"middleware.log$log_api_call$fn__65622.invoke(log.clj:164)"
"middleware.security$add_security_headers$fn__63936.invoke(security.clj:128)"
"middleware.json$wrap_json_body$fn__65327.invoke(json.clj:61)"
"middleware.json$wrap_streamed_json_response$fn__65345.invoke(json.clj:97)"
"middleware.session$bind_current_user$fn__61827$fn__61828.invoke(session.clj:193)"
"middleware.session$do_with_current_user.invokeStatic(session.clj:176)"
"middleware.session$do_with_current_user.invoke(session.clj:170)"
"middleware.session$bind_current_user$fn__61827.invoke(session.clj:192)"
"middleware.session$wrap_current_user_id$fn__61816.invoke(session.clj:161)"
"middleware.session$wrap_session_id$fn__61801.invoke(session.clj:123)"
"middleware.auth$wrap_api_key$fn__63882.invoke(auth.clj:27)"
"middleware.misc$maybe_set_site_url$fn__65649.invoke(misc.clj:56)"
"middleware.misc$bind_user_locale$fn__65652.invoke(misc.clj:72)"
"middleware.misc$add_content_type$fn__65637.invoke(misc.clj:28)"
"middleware.misc$disable_streaming_buffering$fn__65660.invoke(misc.clj:87)")}
```

**Comment by nickhamlin** [comment]
Confirmed also that this only happens when the pulse is trying to use a field where an FK is used to show a different field as the display value instead of the underlying ID

**Comment by flamber** [comment]
@landreville @nickhamlin Did this use to work in previous versions? Slightly related to #10474

**Comment by landreville** [comment]
Unfortunately I hadn't tried to use pulses on previous versions.

**Comment by nickhamlin** [comment]
@flamber We've been using pulses extensively for a while and this hasn't come up before, so, while I can't confirm 100%, my suspicion is that it did used to work in older versions.

In the meantime, I've been working around it either by aggregating the question so that users can at least continue to receive the pulse, click through via the rolled-up number, and then disaggregate to explore the raw data. For situations where we absolutely need the disaggregated data in the pulse, rewriting them in SQL works alright, but does remove the options for more drilldown.

**Comment by adryanm2m** [comment]
I have done some further testing and it only happens if there are 2 or more FK remapped columns in a table. Seems to be the same in earlier versions. Tested in demo datae by remapping Product_ID & User_ID on the orders table

**Comment by adryanm2m** [comment]
This is not a soulution but might help diagnose the problem but increasing the below from 10 to 50 made the problem go away
(def cols-limit
  "Maximum number of columns to render in a Pulse image."
  50)

**Comment by nickhamlin** [comment]
We ran into this issue again today in a bit of a new way. Previously, we'd only seen it when the pulse was showing unaggregated data from a field with a custom FK label. This time though, one of our users got the same error even when that FK field wasn't actually included in the table visual. We were able to get it working by manually deselecting the field from the list in the top bar of the notebook editor (essentially removing it from the SELECT clause entirely under the hood).
